multispecies_whale_detection/dataset.py

```diff
@@ -70,3 +92,257 @@ def new(tfrecord_filepattern: str):
+  """
+  extracted_windows = []
+  extracted_labels = []
+  begin_limit = _audio_duration(features) - context_duration_seconds
+  for _ in range(sample_size):
+    begin = tf.random.uniform([], tf.constant(0.0, tf.float32), begin_limit)
+    window, label = extract_window(features, begin, context_duration_seconds,
+                                   class_names, min_overlap)
+    extracted_windows.append(window)
+    extracted_labels.append(label)
+
+  return tf.stack(extracted_windows), tf.stack(extracted_labels)
+
+
+def sliding_windows(
+    features: Dict[str, tf.Tensor],
+    context_duration_seconds: float,
+    hop_seconds: float,
+    class_names: Sequence[str],
+    min_overlap=DEFAULT_MIN_OVERLAP) -> Tuple[tf.Tensor, tf.Tensor]:
+  """Extracts labeled context windows sliding at a fixed period.
+
+  The number of windows extracted depends on the duration of the audio given.
+  All windows that fit within the bounds of the clip will be returned.
+
+  Args:
+    features: Features of an audio clip as returned by :py:meth:parse_fn.
+    context_duration_seconds: Desired duration of extracted windows. All audio
+      tensors returned will have exactly this duration.
+    hop_seconds: The fixed period at which the slice to extract slides over the
+      audio clip.
+    class_names: A list of all possible values of ANNOTATION_LABEL, also known
+      as the label vocabulary.
+    min_overlap: Lower bound for duration of overlap between an annotation
+      interval and an extracted window for the window to be considered positive
+      for the class in that annotation's label.
+
+  Returns:
+    Audio tensor of shape [sample_size, audio_duration_samples] and labels
+    tensor of float32 class labels of shape [sample_size, len(class_names)].
+  """
+  num_hops = 1 + tf.cast(
+      tf.math.floor(
+          (_audio_duration(features) - context_duration_seconds) / hop_seconds),
+      tf.int32)
+  extracted_windows = tf.TensorArray(tf.float32, size=num_hops)
+  extracted_labels = tf.TensorArray(tf.float32, size=num_hops)
+  begin = 0.0
+  index = 0
+  while begin + context_duration_seconds <= _audio_duration(features):
+    window, label = extract_window(features, begin, context_duration_seconds,
+                                   class_names, min_overlap)
+    extracted_windows = extracted_windows.write(index, window)
+    extracted_labels = extracted_labels.write(index, label)
+    begin += hop_seconds
+    index += 1
+  return extracted_windows.stack(), extracted_labels.stack()
+
+
+def new_window_dataset(
+    tfrecord_filepattern: str,
+    context_duration_seconds: float,
+    class_names: Sequence[str],
+    sample_size: Optional[int] = None,
+    hop_seconds: Optional[float] = None,
+    min_overlap=DEFAULT_MIN_OVERLAP,
+) -> tf.data.Dataset:
+  """Parallel tf.data pipeline to extract windows from TFRecords.
+
+  This is intended for initializing a training set when sample_size is not None
```

Comment:
How would you feel about making the intention explicit with a train/test mode flag (or is_train boolean, which seems more keras-y)? Then we can have arguments train_sample_size and eval_hop_seconds.

Rationale: It's nice to be able to configure options without worrying about mutually exclusive arguments, like new_window_dataset(mode=TRAIN, **config_dict).

Reply:
Point taken on the bad aesthetics of mutually exclusive arguments.

However, I don't find it much prettier to add a third one to disambiguate which of the other two is ignored. I also don't love TRAIN/train_/eval_ naming for assuming the caller's use case, but I think this wasn't your main point.

What I really want is perfectly valid Rust

```
enum Windowing {
  RANDOM(u8),
  SLIDING(f32),
}
```

or C++ std::variant. I don't know whether such a thing exists in Python, do you?

Reply:
I just found [pattern matching](https://www.python.org/dev/peps/pep-0636/#adding-a-ui-matching-objects) is [supposed to come to Python 3.10](https://mail.python.org/archives/list/python-committers@python.org/message/SQC2FTLFV5A7DV7RCEAR2I2IKJKGK7W3/). For now, I'll [use isinstance](http://blog.ezyang.com/2020/10/idiomatic-algebraic-data-types-in-python-with-dataclasses-and-union/).

Reply:
I guess object-oriented (strategy pattern) is the Pythonic way to do this, unless we really needed "windowing" to be a data object, but there's no real need for that yet.